### PR TITLE
[DNM] Add Donation::cancel, Donation::getDecodedData and Donation::encodeAndSetData

### PR DESCRIPTION
**Pull request by JeroenDeDauw**

### CODE REVIEW

**Comment by wmde-manicki** [comment]
+1 but @KaiNissen should also have a look if there is something more to be set on cancellation.

Also as discussed in https://github.com/wmde/FundraisingStore/issues/36 I don't think `dt_del` and `status` are needed as keys in the data field but I might be wrong here.

**Comment by JeroenDeDauw** [comment]
> Also as discussed in #36 I don't think dt_del and status are needed as keys in the data field but I might be wrong here.

FYI: This is copied from the old code (which ofc does not mean it is actually needed behaviour)

**Comment by JeroenDeDauw** [comment]
Actually thinking now it's probably better to not have cancel here at all and just put it in the domain object.

**Comment by JeroenDeDauw** [comment]
Needs rebase

File: src/Entities/Donation.php [hunk 1]
```diff
@@ -692,7 +689,41 @@ public function getEntryType( $mode = null ) {
...
+	 * @since 2.0
+	 */
+	public function cancel() {
+		$this->dtDel = date( 'Y-m-d H:i:s' );
+		$this->status = 'D';
+
+		$data = $this->getDecodedData();
+
+		$data['dt_del'] = $this->dtDel;
+		$data['status'] = $this->status;
```
Comment (JeroenDeDauw) [inline src/Entities/Donation.php]: Can we drop this status field? The entity itself already has one. And if we can't drop it, is there a problem with on every save of the entity to set this status to the one of the entities field?

File: src/Entities/Donation.php [hunk 2]
```diff
@@ -692,7 +689,41 @@ public function getEntryType( $mode = null ) {
...
+		$data['utoken'] = '';
```
Comment (JeroenDeDauw) [inline src/Entities/Donation.php]: Is this token reset really needed? The user will already be unable to do updates as we check the status no? Ping @KaiNissen @gbirke

File: src/Entities/Donation.php (hunk 1 above)
Comment (KaiNissen) [inline src/Entities/Donation.php]: I'm quite sure, we can drop this.

File: src/Entities/Donation.php (hunk 2 above)
Comment (KaiNissen) [inline src/Entities/Donation.php]: As far as I remember, the old application never reset the token. I don't think we should/need to do that.

Comment (JeroenDeDauw) [inline src/Entities/Donation.php]: Uh - I think I put it here because I saw the old code doing it :)

**Pull request closed without merging**